src/Dialect/ONNX/ShapeInference/ONNXShapeHelper.cpp

```diff
@@ -399,44 +400,57 @@ LogicalResult ONNXGenericPoolShapeHelper<OP_TYPE, OP_ADAPTOR>::computeShape(
   return success();
 }
 
-/// Check if the given inferredShape is better than the existing shape of val.
-bool isInferredShapeBetter(ArrayRef<int64_t> inferredShape, Value val) {
+/// Use the val's shape to improve the given shape.
+/// - If the val is unranked, no update.
+/// - If a dim is static in the val's shape but unknown (-1) in the given shape,
+/// update it in the given shape to the static value.
+void tryImproveInferredShape(
+    SmallVectorImpl<int64_t> &inferredShape, Value val) {
   // Val is unranked. Any inferred shape is better.
   if (!hasShapeAndRank(val))
-    return true;
-
-  ArrayRef<int64_t> existingShape = val.getType().cast<ShapedType>().getShape();
-
-  // Only get the infered shape if unknown dims in existingShape get static
-  // values in inferredShape.
-  for (unsigned i = 0; i < existingShape.size(); ++i) {
-    int64_t inferedDim = inferredShape[i];
-    int64_t existingDim = existingShape[i];
-    // existingDim is static but inferedDim is unknown: not good.
-    if ((existingDim != -1) && (inferedDim == -1))
-      return false;
+    return;
+
+  ArrayRef<int64_t> existingShape = getShape(val.getType());
+  assert(inferredShape.size() == existingShape.size() &&
+         "Inferred shape and existing shape are inconsistent in the number of "
+         "elements");
+
+  for (unsigned i = 0; i < inferredShape.size(); ++i) {
+    // existingDim is static but inferedDim is unknown: update the inferredDim.
+    if ((existingShape[i] != -1) && (inferredShape[i] == -1))
+      inferredShape[i] = existingShape[i];
     // inferedDim is different from existingDim. Believe in existingDim.
-    if ((existingDim != -1) && (inferedDim != -1) &&
-        (existingDim != inferedDim))
-      return false;
+    if ((existingShape[i] != -1) && (inferredShape[i] != -1) &&
+        (existingShape[i] != inferredShape[i])) {
+      // Warning for users.
+      llvm::outs() << "Shape inference: the inferred dim (" << inferredShape[i]
```

Comment:
isn't there an error stream? It would also be good to preface it with "Warning:"